python/pyarrow/tests/test_array.py

```diff
@@ -1046,6 +1046,21 @@ def test_array_conversions_no_sentinel_values():
     assert arr3.null_count == 0
 
 
+def test_binary_string_pandas_null_sentinels():
+    def _check_case(ty):
+        with pytest.raises(TypeError):
+            # NaN is a float by default
+            pa.array(['string', np.nan], type=ty)
+
+        arr = pa.array(['string', np.nan], type=ty, from_pandas=True)
+        expected = pa.array(['string', None], type=ty)
+        assert arr.equals(expected)
+
+    _check_case(None)
+    _check_case('binary')
+    _check_case('utf8')
+
```

Comment:
It might be worth to add some additional test cases. Some things that worked on 0.14.1 (but that should fail):

```
In [17]: pa.array([pd.Timestamp("2012"), np.nan]) 
Out[17]: 
<pyarrow.lib.TimestampArray object at 0x7febe0090888>
[
  2012-01-01 00:00:00.000000,
  null
]

In [19]: import datetime 

In [20]: pa.array([datetime.date(2012, 1, 1), np.nan]) 
Out[20]: 
<pyarrow.lib.Date32Array object at 0x7febe00700f8>
[
  2012-01-01,
  null
]

In [26]: pa.array([False, True, np.nan]) 
Out[26]: 
<pyarrow.lib.BooleanArray object at 0x7febe00a5e08>
[
  false,
  true,
  null
]
```

The above now all work with this branch (and struct was already working).

One thing I noticed that does not yet work correctly is pd.NaT for the datetime case:

```
In [6]: pa.array([pd.Timestamp("2012-01-01"), pd.NaT])                                                                                                                                                             
Out[6]: 
<pyarrow.lib.TimestampArray object at 0x7f46c8368780>
[
  2012-01-01 00:00:00.000000,
  0001-01-01 00:00:00.000000
]
```

where `pd.NaT` is converted to "0001-01-01", which is strange, as that does not even correspond with the integer value of pd.NaT. I opened https://issues.apache.org/jira/browse/ARROW-6305 for this.